lib/kafkaClient.js

```diff
@@ -695,6 +726,29 @@ KafkaClient.prototype.clearCallbackQueue = function (socket, error) {
   delete this.cbqueue[socketId];
 };
 
+// Sends a version 1 metadata request which includes controller node in response
+KafkaClient.prototype.loadMetadataForTopicsV2 = function (topics, cb) {
```

Comment:
We shouldn't continue to create these V{x} version of api calls. It creates unnecessary burden of managing kafka version protocol on the user:

- Override Client `loadMetadataForTopics`
- ensure the broker is ready  and use `getSupportedForRequestType` to get the appropriate version of encoder/decoder  (see example in `sendRequest`)

Reply:
Makes sense - my one concern with this is is that the between api version 0 and >=1 the semantics have changed with regards to the topics parameter.

In version 0 an empty array returns all topics. In version >= 1, a null value (-1) is used to fetch all topics. An empty array returns no topic metadata [1]. This is potentially trouble for clients that use this. I guess this could be put under the next major version umbrella, thus allowing this breaking change. But there are also plenty of calls from inside the lib relying on this (and I think doing something like converting [] -> null would be a bit of a lie for clients).

A suggestion could be:
 - Override it, and make it work with getSupportedForRequestType, which includes the different behaviour depending on version, thus a breaking change similar to createTopics.
 - It is up to clients to be aware of this - basically if they use a version >= 0.10 "null" is now used for fetching all topics.
 - Allow internal use to force version 0 for now to avoid manipulating the input parameter in order to keep all existing functionality that works with < 0.10, running. 

[1] (https://kafka.apache.org/protocol#The_Messages_Metadata).

Reply:
Seems like an ugly change from a users perspective since it puts the burden on the user again to update the behavior of their app depending on the version they're connecting to and would make migrations painful. I think this is a case where we should keep kafka node's behavior consistent and abstract away the ugliness of the kafka  protocol. Calling `loadMetadataForTopics` with empty array should always return all topics.

I always thought fetching all topic metadata seemed inefficient when we're just to get the broker metadata. So perhaps we can have new method called `loadMetadata` that will always pass [] for topic?

What do you think?

Reply:
So the following changes:

- Change existing `loadMetaDataForTopics` so that if the broker supports api version => 1, then convert an empty array to null value to fetch all topics
- Add new `loadMetadata` that doesen't request for topics (by always passing [] for topic, I assume you mean fetch no topics in this case?). This can then be used in other cases where we just need broker metadata.

This sounds good to me - if you agree i'll try and get all the suggested changes in the PR in sometime this week.

Reply:
sounds great @jlandersen thanks. 👍